File: architecture-skills/design-patterns/examples/mediator/mediator.swift
// Mediator pattern in Swift.
//
// The chat room mediates between users via a protocol. Users hold a weak
// reference to the room to avoid a retain cycle (the room strongly holds
// its users). NotificationCenter is the stdlib's built-in mediator for
// broadcast-style decoupling.

import Foundation

protocol ChatMediator: AnyObject {
    func broadcast(from sender: User, text: String)
    func direct(from sender: User, to recipient: String, text: String)
}

final class User {
    let name: String
    weak var room: ChatMediator?

    init(name: String) {
        self.name = name
    }

    func say(_ text: String) {
        room?.broadcast(from: self, text: text)
    }

    func dm(_ recipient: String, _ text: String) {
        room?.direct(from: self, to: recipient, text: text)
    }

    func receive(from sender: String, text: String) {
        print("[\(name)] \(sender): \(text)")
    }
}

final class ChatRoom: ChatMediator {
    private var users: [User] = []

    func register(_ user: User) {
        users.append(user)
        user.room = self
    }

    func broadcast(from sender: User, text: String) {
        for user in users where user !== sender {
            user.receive(from: sender.name, text: text)
        }
    }

    func direct(from sender: User, to recipient: String, text: String) {
        guard let user = users.first(where: { $0.name == recipient }) else {
            sender.receive(from: "room", text: "no such user: \(recipient)")
            return
        }
        user.receive(from: sender.name, text: "(dm) \(text)")
    }
}

let room = ChatRoom()
let ada = User(name: "Ada")
let grace = User(name: "Grace")
let alan = User(name: "Alan")
for user in [ada, grace, alan] {
    room.register(user)
}

ada.say("Hello, everyone!")
grace.dm("Ada", "Hi Ada, got a minute?")
alan.dm("Linus", "ping")

File: architecture-skills/design-patterns/examples/mediator/mediator.rs
// Mediator pattern in Rust.
//
// Cross-references between colleagues fight the borrow checker, so the
// idiomatic shape is to let the mediator OWN the colleagues and address
// them by name/id. Callers talk to the room, never to another user.
// (The GoF form with shared colleagues needs Rc<RefCell<User>>.)

use std::collections::BTreeMap;

struct User {
    name: String,
}

impl User {
    fn receive(&self, sender: &str, text: &str) {
        println!("[{}] {}: {}", self.name, sender, text);
    }
}

struct ChatRoom {
    users: BTreeMap<String, User>,
}

impl ChatRoom {
    fn new() -> Self {
        Self {
            users: BTreeMap::new(),
        }
    }

    fn register(&mut self, name: &str) {
        self.users.insert(
            name.to_string(),
            User {
                name: name.to_string(),
            },
        );
    }

    fn broadcast(&self, sender: &str, text: &str) {
        for user in self.users.values() {
            if user.name != sender {
                user.receive(sender, text);
            }
        }
    }

    fn direct(&self, sender: &str, recipient: &str, text: &str) {
        match self.users.get(recipient) {
            Some(user) => user.receive(sender, &format!("(dm) {text}")),
            None => {
                if let Some(user) = self.users.get(sender) {
                    user.receive("room", &format!("no such user: {recipient}"));
                }
            }
        }
    }
}

fn main() {
    let mut room = ChatRoom::new();
    room.register("Ada");
    room.register("Grace");
    room.register("Alan");

    room.broadcast("Ada", "Hello, everyone!");
    room.direct("Grace", "Ada", "Hi Ada, got a minute?");
    room.direct("Alan", "Linus", "ping");
}
